Check mathType only if not Float32

RomeoV opened the pull request:
**Summary**
This fixes a bug where the convolution algorithm selector filters out valid algorithms.

**Describe the bug**

I am running relatively simple Deep Learning Vision Models, like ResNet (and custom models) on my nvidia gpu.
Recently, I updated my dependencies and have started to see the following error, along with a huge performance regression:
```
┌ Warning: No valid algorithm found, probably bad params for convolution.
└ @ cuDNN ~/.julia/packages/cuDNN/3J08S/src/convolution.jl:280
```

Here's a MWE reproducing the problem:
```julia
import Pkg; Pkg.activate(temp=true); Pkg.add(["Flux", "Metalhead", "CUDA"])
using Flux, Metalhead, CUDA
model = Metalhead.ResNet(18; pretrain=false) |> gpu;
x = CUDA.rand(Float32, 32, 32, 3, 7);
model(x) |> sum
```

As you can see, this is the "most basic" form of instantiating and evaluating a Flux-based model.
The error has also come up from models that are not from Metalhead, but are self-defined from basic building blocks instead.

Here's some information on my system:
```julia
julia> versioninfo()
Julia Version 1.9.0
Commit 8e630552924 (2023-05-07 11:25 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

```bash
[romeo@Romeo-P1 VAEModels.jl]$ nvidia-smi
Thu Jun  8 04:01:35 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.41.03              Driver Version: 530.41.03    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA RTX A2000 Laptop GPU     Off| 00000000:01:00.0 Off |                  N/A |
| N/A   55C    P8                8W /  N/A|      2MiB /  4096MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
```

```
julia> CUDA.versioninfo()
CUDA runtime 12.1, artifact installation
CUDA driver 12.1
NVIDIA driver 530.41.3

CUDA libraries: 
- CUBLAS: 12.1.3
- CURAND: 10.3.2
- CUFFT: 11.0.2
- CUSOLVER: 11.4.5
- CUSPARSE: 12.1.0
- CUPTI: 18.0.0
- NVML: 12.0.0+530.41.3

Julia packages: 
- CUDA.jl: 4.3.2
- CUDA_Driver_jll: 0.5.0+1
- CUDA_Runtime_jll: 0.6.0+0
- CUDA_Runtime_Discovery: 0.2.2

Toolchain:
- Julia: 1.9.0
- LLVM: 14.0.6
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5
- Device capability support: sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80, sm_86

1 device:
  0: NVIDIA RTX A2000 Laptop GPU (sm_86, 2.996 GiB / 4.000 GiB available)

```

RomeoV commented:
This also happens when launching the above MWE with `julia --pkgimages=no mwe.jl`.

RomeoV commented:
I've managed to reduce the model to a much smaller model that shows the same behavior.
If we replace the model in the MWE with the following
```julia
model = Chain(Conv((5, 5), 3=>32, leakyrelu; pad=SamePad()),
              Conv((1, 1), 32=>64, identity),
              MeanPool((2, 2)),
              Conv((1, 1), 64=>128, identity),
              ) |> gpu
```
**For the model above, the same warning / performance regression occurs.**

However, either of the following models do not have this problem:
```julia
# remove `SamePad()`
model = Chain(Conv((5, 5), 3=>32, leakyrelu),
              Conv((1, 1), 32=>64, identity),
              MeanPool((2, 2)),
              Conv((1, 1), 64=>128, identity),
              ) |> gpu
# or
# remove `MeanPool((2, 2))`
model = Chain(Conv((5, 5), 3=>32, leakyrelu; pad=SamePad()),
              Conv((1, 1), 32=>64, identity),
              Conv((1, 1), 64=>128, identity),
              ) |> gpu
```

Tbh, I'm pretty stumped as to what's going on...

RomeoV commented:
Here's another model that does not have the problem:
```julia
# start with the non-working model from above, but reduce the final output dimension from 128 to 64
model = Chain(Conv((5, 5), 3=>32, leakyrelu; pad=SamePad()),
              Conv((1, 1), 32=>64, identity),
              MeanPool((2, 2)),
              Conv((1, 1), 64=>64, identity),
              ) |> gpu
```
This does not show the warning / performance regression.

EDIT: This doesn't work anymore with larger batch sizes. I.e. in the MWE I use batch size 7, and this model (with reduced parameters) works fine. With a larger batchsize of 64, I get the same problem as above again, i.e. "No valid algorithm found..." and the performance regression.

RomeoV commented:
It seem that this problem has been introduced in version 4.3 of CUDA.
Specifying the CUDA version in the MWE shows the problem in 4.3.0, but not in 4.2.
However, I'm not sure yet if it's just that the warning is not shown, or if the performance regression is also gone.

RomeoV commented:
No idea what happened, but the problem seemed to have fixed itself. Closing for now.

RomeoV commented:
Nvm, I just encountered it again, but only on larger models, not on the toy problems above...
Needs further investigation, but I'm a bit strapped for time right now.

I double checked though that this happens on `master` and the latest tag `v4.3.2`, but not on `v4.2.0`.
If I find a bit more time I can bisect the commits, not today though.

maleadt commented:
Maybe it's a bug in CUDNN. I'll update the JLLs.

maleadt commented:
Please test with https://github.com/JuliaGPU/CUDA.jl/pull/1959

RomeoV commented:
Unfortunately the ~~error~~ warning persists also with #1959.

RomeoV commented:
I tried bisecting and was able to figure out that the regression happened between `v4.2.0` and `v4.3.0`. Unfortunately, I can't seem to (pre)compile `cuDNN` for any of the in-between commits, as they all error with some message related to `@retry_reclaim` missing (even after the the #1906 merge). However, the tagged commits build fine.

maleadt commented:
You need to `dev` both CUDA and cuDNN.

RomeoV commented:
Thanks! Git bisect yielded that the problem comes from b53b13039 by @ToucheSir , which makes sense since it concerns the algorithm selection.
The line is probably [this one](https://github.com/JuliaGPU/CUDA.jl/blob/b53b1303940888dfe0e86bc92a36db6af55616ba/lib/cudnn/src/convolution.jl#L275), although I can't see the problem with it right away.

Happy to hear some thoughts on why this commit was necessary?
Following the cuda documentation ([here for convenience](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnFindConvolutionBackwardFilterAlgorithmEx)) linked in the reverted commit suggests to me that the CUDNN_DEFAULT_MATH should be fine usually. However, prevent cuda from using it with this commit.
Happy to hear any thoughts @ToucheSir.

RomeoV committed:
Revert "cuDNN: only choose conv algorithms if they match descriptor mathType (#1917)"

This reverts commit b53b1303940888dfe0e86bc92a36db6af55616ba.
Fixes #1943.

RomeoV commented:
I also opened a PR reverting the commit, although I didn't mean to turn the issue into a PR right away...

ToucheSir commented:
> Happy to hear some thoughts on why this commit was necessary?

If you follow the link back to the [PR](https://github.com/JuliaGPU/CUDA.jl/pull/1917), you'll see that not having this check was causing some types of conv operations to completely fail. That in itself was a hard regression, whereas I see this as more of a "soft" regression since it appears to manifest only when you're close to the memory limit on your GPU.

One frustrating thing with that PR is that I could find no documentation on the cuDNN side which said conv ops could fail if the descriptor math mode doesn't match the supported ones by the algorithm. If you can find documentation for that or figure out what the rule is, I'd be happy to help with drafting and reviewing a PR which refines the logic.

A quick test you can try is to set `CUDA.math_mode!(CUDA.PEDANTIC_MATH)` before running any GPU code. That should activate https://github.com/JuliaGPU/CUDA.jl/blob/3321fc8aa53fee2ac39783ad1119af665545750b/lib/cudnn/src/cuDNN.jl#L52, which in turn should ensure the conv descriptor and algorithm supported math modes always match because `CUDNN_DEFAULT_MATH` is universal the fallback.

RomeoV commented:
Indeed turning on `CUDA.math_mode!(CUDA.PEDANTIC_MATH)` fixes the problem. I'll have to look more deeply to understand what's going on - I'm not super clear on what the math type is supposed to imply.
At least for now, this is a fix for anybody running across this problem.

ToucheSir commented:
After a lot of digging through PyTorch code trying to find out how they handle this, I found https://github.com/pytorch/pytorch/blame/0d653730ce4314c4c48da0e336c1e1c6259ada28/aten/src/ATen/native/cudnn/Conv_v8.cpp#L131-L134 which looks promising. Translated back into CUDA.jl terms, I think we should try skipping the `mathType` check if the input eltype is `Float32`. Care to give that a try @RomeoV?

RomeoV committed:
Don't check mathType if eltype is float32

RomeoV commented:
Thanks @ToucheSir, this seems to be doing the job! In my tests the problem seems solved now.
I referenced your comment above in the code, let me know if you would prefer that another way.

ToucheSir commented on lib/cudnn/src/convolution.jl:
```diff
@@ -264,15 +264,20 @@ end
     mathType = Ref{cudnnMathType_t}(CUDNN_DEFAULT_MATH)
     cudnnGetConvolutionMathType(convDesc, mathType)
+    skipMathTypeCheck = begin  # See https://github.com/JuliaGPU/CUDA.jl/pull/1943#issuecomment-1605267932
+        dtype, _ = cudnnGetTensorDescriptor(tensorDesc)
```
Per https://buildkite.com/julialang/cuda-dot-jl/builds/4082#0189195a-1428-4cf5-84b7-55376287aee3/275-443 this needs to handle different types of descriptors. Other than that the changes look fine to me.

RomeoV committed:
Change filterDescriptor to tensorDescriptor

ToucheSir commented on lib/cudnn/src/convolution.jl:
```diff
@@ -264,15 +264,20 @@ end
+function cudnnConvolutionAlgoPerfChoose(convDesc, tensorDesc, perfResults, n)
     mathType = Ref{cudnnMathType_t}(CUDNN_DEFAULT_MATH)
     cudnnGetConvolutionMathType(convDesc, mathType)
+    skipMathTypeCheck = begin  # See https://github.com/JuliaGPU/CUDA.jl/pull/1943#issuecomment-1605267932
```
```suggestion
    skipMathTypeCheck = let  # See https://github.com/JuliaGPU/CUDA.jl/pull/1943#issuecomment-1605267932
```

RomeoV committed:
begin -> let

Recall that unlike `begin`,  `let` doesn't leak defined variables into outside scope.

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>
